sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala

```diff
@@ -163,6 +163,155 @@ class CollationStringExpressionsSuite
+        SubstringTestCase("select right('AaAaAaAa000000' collate " + c + ", 3)", c, Row("000")),
+        SubstringTestCase("select substr('' collate " + c + ", 1, 1)", c, Row("")),
+        SubstringTestCase("select left('' collate " + c + ", 1)", c, Row("")),
+        SubstringTestCase("select right('' collate " + c + ", 1)", c, Row("")),
+        SubstringTestCase("select left('ghi' collate " + c + ", 1)", c, Row("g"))
+      )
+    )
+
+    checks.foreach { check =>
+      // Result & data type
+      checkAnswer(sql(check.query), check.result)
+      assert(sql(check.query).schema.fields.head.dataType.sameType(StringType(check.collation)))
+    }
+  }
+
+  test("Support Left/Right/Substr with explicit improper collation") {
+    case class SubstringTestCase(query: String, collation: String, result: Row)
+    val checks = Seq("utf8_binary_lcase", "utf8_binary", "unicode", "unicode_ci").flatMap(
+      c => Seq(
+        SubstringTestCase("select left(null collate " + c + ", 1)", c, Row(null)),
+        SubstringTestCase("select right(null collate " + c + ", 1)", c, Row(null)),
+        SubstringTestCase("select substr(null collate " + c + ", 1)", c, Row(null)),
+        SubstringTestCase("select substr(null collate " + c + ", 1, 1)", c, Row(null)),
+        SubstringTestCase("select left('' collate " + c + ", null)", c, Row(null)),
+        SubstringTestCase("select right('' collate " + c + ", null)", c, Row(null)),
+        SubstringTestCase("select substr('' collate " + c + ", null)", c, Row(null)),
+        SubstringTestCase("select substr('' collate " + c + ", null, null)", c, Row(null))
+      )
+    )
+    checks.foreach(check => {
+      // Result & data type
+      checkAnswer(sql(check.query), check.result)
+      assert(sql(check.query).schema.fields.head.dataType.sameType(StringType(check.collation)))
+    })
+  }
+
+  test("Support Left/Right/Substr with explicit improper length & position") {
+    case class SubstringTestCase(query: String, collation: String, result: Row)
+    val checks = Seq("utf8_binary_lcase", "utf8_binary", "unicode", "unicode_ci").flatMap(
+      c => Seq(
+        SubstringTestCase("select left(' a世a ' collate " + c + ", '3')", c, Row(" a世")), // scalastyle:ignore
+        SubstringTestCase("select right(' a世a ' collate " + c + ", '3')", c, Row("世a ")), // scalastyle:ignore
+        SubstringTestCase("select right('' collate " + c + ", null)", c, Row(null)),
+        SubstringTestCase("select substr('' collate " + c + ", null)", c, Row(null)),
+        SubstringTestCase("select substr('' collate " + c + ", null, null)", c, Row(null)),
+        SubstringTestCase("select left('' collate " + c + ", null)", c, Row(null))
+      )
+    )
+    checks.foreach(check => {
+      // Result & data type
+      checkAnswer(sql(check.query), check.result)
+      assert(sql(check.query).schema.fields.head.dataType.sameType(StringType(check.collation)))
+    })
+  }
```

Comment:
improper length & position in context of collations should only matter with variable byte length characters, so make sure to use such examples in `test("Support Left/Right/Substr with collation")` to keep this test particular suite to a minimum